src/main/kotlin/ai/whylabs/services/whylogs/kafka/KotlinConsumer.kt

```diff
@@ -0,0 +1,122 @@
+    val consumerConfig = mapOf(
+        ConsumerConfig.BOOTSTRAP_SERVERS_CONFIG to config.bootstrapServers,
+        ConsumerConfig.GROUP_ID_CONFIG to config.groupId,
+        ConsumerConfig.KEY_DESERIALIZER_CLASS_CONFIG to "org.apache.kafka.common.serialization.StringDeserializer",
+        ConsumerConfig.VALUE_DESERIALIZER_CLASS_CONFIG to "org.apache.kafka.common.serialization.StringDeserializer",
+    )
+
+    log.info("Setting up kafka consumer with config $consumerConfig")
+    val consumer = KafkaConsumer<String, String>(consumerConfig)
+    consumer.subscribe(config.topics)
+    return consumer
+}
+
+data class KotlinConsumerConfig<K, V>(
+    val consumerCount: Int = 1,
+    val topics: List<String>,
+    val bootstrapServers: List<String>,
+    val groupId: String,
+)
+
+private val mapper = jacksonObjectMapper()
+
+data class KafkaRecord<K, V>(
+    val value: V,
+    val key: K?,
+    val rawRecord: ConsumerRecord<String, String>
+)
+
+class KotlinConsumer<K, V>(
+    private val config: KotlinConsumerConfig<K, V>,
+    private val valueType: TypeReference<V>,
+    private val keyType: TypeReference<K>,
+    private val process: suspend (record: KafkaRecord<K, V>) -> Boolean
+) :
+    CoroutineScope {
+
+    companion object {
+        inline fun <reified K, reified V> consumer(
+            config: KotlinConsumerConfig<K, V>,
+            noinline processor: suspend (record: KafkaRecord<K, V>) -> Boolean
+        ): KotlinConsumer<K, V> {
+            return KotlinConsumer(config, object : TypeReference<V>() {}, object : TypeReference<K>() {}, processor)
+        }
+    }
+
+    init {
+        Runtime.getRuntime().addShutdownHook(Thread { stop() })
+    }
+
+    private val supervisorJob = SupervisorJob()
+    override val coroutineContext: CoroutineContext
+        get() = Dispatchers.IO + supervisorJob
+
+    fun stop() {
+        supervisorJob.cancel()
+    }
+
+    fun start() = launch {
+        log.info("Creating ${config.consumerCount} consumers with dedicated threads.")
+        repeat(config.consumerCount) {
+            launchWorker()
+        }
+    }
+
+    // TODO need to make it reconnect if the cluster goes down
+    private fun CoroutineScope.launchWorker() = launch {
+        val consumer = getConsumer<K, V>(config)
+
+        repeatUntilCancelled(log) {
+            log.debug("Polling for messages")
+            val records = consumer.poll(Duration.ofSeconds(10))
+
+            if (records.count() == 0) {
+                // do nothing
+                log.debug("Got no records to process.")
+            } else {
+                log.info("Got ${records.count()} records to process.")
+                records.forEach forEach@{ record ->
+                    // TODO how should you handle partial failures within messages batches? We prob don't have this problem
+                    // But I should know the answer.
+                    // TODO how should we handle errors here? What if the message is nonsense?

```

Comment:
feel like my help here to you is close to 0 haha but this might be something to look at: https://www.confluent.io/blog/error-handling-patterns-in-kafka/

Reply:
Nice, that's helpful. I think the only one we would consider doing is a separate queue to dump the error messages. Retries won't change anything in our case so a full DLQ wouldn't make sense. We would just give people an option in the future to see what isn't being included in their whylogs profiles, and the most we can do then is try to run whylogs on those message to see what's wrong.